File: core/orchestrator.py
"""Orchestration logic for theme song retrieval."""

import logging
from pathlib import Path
from typing import List, Optional
from rich.console import Console
from rich.table import Table
from scrapers.base import ThemeScraper
from scrapers.tv_tunes import TelevisionTunesScraper
from scrapers.anime_themes import AnimeThemesScraper
from scrapers.themes_moe import ThemesMoeScraper
from scrapers.youtube import YoutubeScraper
from core.config import Config
from core.rate_limiter import RateLimiter
from core.logging_utils import StructuredLogger
from core.utils import get_file_size_formatted, get_audio_duration


class CriticalError(Exception):
    """Exception raised for critical errors that should stop execution."""


class Orchestrator:
    """Orchestrates theme song retrieval across multiple sources."""

    def __init__(
        self,
        console: Console,
        force: bool = False,
        dry_run: bool = False,
        verbose: bool = False,
        timeout: int = Config.DEFAULT_TIMEOUT_SEC
    ):
        """
        Initialize the orchestrator.

        Args:
            console: Rich console for output
            force: If True, overwrite existing theme files
            dry_run: If True, simulate operations without downloading
            verbose: If True, enable debug logging
            timeout: Network timeout in seconds
        """
        self.console = console
        self.force = force
        self.dry_run = dry_run
        self.verbose = verbose
        self.timeout = timeout
        self.scrapers: List[ThemeScraper] = []
        self.logger = logging.getLogger(__name__)
        self.structured_logger = StructuredLogger(__name__)
        self.rate_limiter = RateLimiter()
        self.results = {
            "success": 0,
            "skipped": 0,
            "failed": 0
        }

        # Initialize scrapers in priority order
        self._initialize_default_scrapers()

    def _initialize_default_scrapers(self) -> None:
        """
        Initialize scrapers in priority order.

        Priority order:
        1. TelevisionTunes (best for TV shows)
        2. AnimeThemes (best for anime)
        3. Themes.moe (additional anime source)
        4. YouTube (fallback for everything)
        """
        self.scrapers = [
            TelevisionTunesScraper(self.console, self.verbose, self.timeout),
            AnimeThemesScraper(self.console, self.verbose, self.timeout),
            ThemesMoeScraper(self.console, self.verbose, self.timeout),
            YoutubeScraper(self.console, self.verbose)
        ]

    def add_scraper(self, scraper: ThemeScraper) -> None:
        """
        Add a scraper to the orchestrator.

        Args:
            scraper: ThemeScraper instance to add
        """
        self.scrapers.append(scraper)

    def process_directory(self, input_dir: Path) -> None:
        """
        Process all series folders in the input directory.

        Args:
            input_dir: Root directory containing series folders

        Raises:
            CriticalError: If a critical error occurs that should stop execution
        """
        # Validate input directory
        if not input_dir.exists():
            raise CriticalError(f"Input directory does not exist: {input_dir}")

        if not input_dir.is_dir():
            raise CriticalError(f"Input path is not a directory: {input_dir}")

        self.logger.info(f"Scanning directory: {input_dir}")
        series_folders = self._scan_directory(input_dir)

        if not series_folders:
            self.console.print("[yellow]No series folders found[/]")
            self.logger.warning("No series folders found in input directory")
            return

        total_folders = len(series_folders)
        self.console.print(f"[bold dodger_blue1]Found {total_folders} series folders[/bold dodger_blue1]\n")
        self.logger.info(f"Found {total_folders} series folders to process")

        # Process folders sequentially (simpler and safer than async)
        for index, folder in enumerate(series_folders, start=1):
            self.console.print(f"\n[bold magenta]━━━ Folder {index}/{total_folders} ━━━[/bold magenta]")
            self.logger.info(f"Processing folder {index}/{total_folders}: {folder.name}")
            try:
                self.process_show(folder)
            except CriticalError:
                # Re-raise critical errors to stop execution
                raise
            except Exception as exc:
                # Log non-critical errors and continue processing
                self.logger.error(
                    f"Unexpected error processing {folder.name}: {str(exc)}",
                    exc_info=True
                )
                self.console.print(
                    f"[red]ERROR[/] Unexpected error processing {folder.name}: {str(exc)}"
                )
                if self.verbose:
                    import traceback
                    self.console.print(traceback.format_exc())
                self.results["failed"] += 1

        self.logger.info(
            f"Processing complete - Success: {self.results['success']}, "
            f"Skipped: {self.results['skipped']}, Failed: {self.results['failed']}"
        )
        self.display_summary()


    def _scan_directory(self, input_dir: Path) -> List[Path]:
        """
        Scan directory for series folders.

        Args:
            input_dir: Directory to scan

        Returns:
            List of Path objects for each series folder
        """
        series_folders = []

        try:
            for item in input_dir.iterdir():
                if item.is_dir():
                    series_folders.append(item)
        except PermissionError:
            self.console.print(
                f"[yellow]Warning: Permission denied accessing {input_dir}[/]"
            )
        except OSError as e:
            self.console.print(
                f"[yellow]Warning: Error accessing {input_dir}: {str(e)}[/]"
            )

        return series_folders

    def _extract_show_name(self, folder: Path) -> str:
        """
        Extract show name from folder path.

        Args:
            folder: Path to series folder

        Returns:
            Show name (folder's base name)
        """
        return folder.name

    def _find_existing_theme(self, folder: Path) -> Optional[Path]:
        """
        Check if a theme file already exists in the folder.

        Checks for theme files with extensions from Config.THEME_EXTENSIONS

        Args:
            folder: Path to series folder

        Returns:
            Path to existing theme file, or None if not found
        """
        for ext in Config.THEME_EXTENSIONS:
            theme_file = folder / f"theme{ext}"
            if theme_file.exists():
                return theme_file
        return None


    def process_show(self, folder: Path) -> None:
        """
        Process a single show folder.

        Args:
            folder: Path to series folder

        Raises:
            CriticalError: If a critical error occurs that should stop execution
        """
        show_name = self._extract_show_name(folder)
        theme_file = folder / "theme.mp3"

        # Check for permission errors accessing the folder
        try:
            # Test if we can access the folder
            folder.exists()
            list(folder.iterdir())
        except PermissionError:
            self.console.print(
                f"[dim yellow]⊘ SKIPPED[/dim yellow] {show_name} [dim]- Permission denied[/dim]"
            )
            self.results["skipped"] += 1
            return
        except OSError as exc:
            self.console.print(
                f"[dim yellow]⊘ SKIPPED[/dim yellow] {show_name} [dim]- Cannot access folder: {str(exc)}[/dim]"
            )
            self.results["skipped"] += 1
            return

        # Check if theme already exists
        existing_theme = self._find_existing_theme(folder)

        if existing_theme and not self.force:
            self.console.print(f"[dim yellow]⊘ SKIPPED[/dim yellow] {show_name} [dim]- File exists[/dim]")
            self.logger.info(f"Skipped '{show_name}' - theme file already exists")
            self.results["skipped"] += 1
            return

        if self.dry_run:
            self.console.print(f"[bright_blue]⚡ DRY RUN[/bright_blue] Would process: {show_name}")
            self.logger.info(f"Dry run - would process '{show_name}'")
            return

        # If force mode is enabled and a theme exists, delete it before downloading
        if self.force and existing_theme:
            try:
                existing_theme.unlink()
                if self.verbose:
                    self.console.print(f"  Deleted existing theme: {existing_theme.name}")
                self.logger.info(f"Deleted existing theme for '{show_name}' (force mode)")
            except OSError as exc:
                self.console.print(
                    f"[yellow]Warning: Could not delete existing theme: {str(exc)}[/]"
                )
                self.logger.warning(f"Could not delete existing theme: {str(exc)}")

        # Try each scraper in order (waterfall approach)
        self.console.print(f"\n[bold white]Processing:[/bold white] [dodger_blue1]{show_name}[/dodger_blue1]")

        for scraper in self.scrapers:
            source_name = scraper.get_source_name()
            self.console.print(f"  [dim]Trying[/dim] {source_name}...", end="")
            self.logger.info(f"Attempting source: {source_name} for '{show_name}'")

            # Apply rate limiting before attempting scraper
            self.rate_limiter.wait(source_name)

            try:
                if scraper.search_and_download(show_name, theme_file):
                    self.console.print(f" [bold green]✓[/bold green]")
                    
                    # Get file metadata
                    file_size = get_file_size_formatted(theme_file)
                    duration = get_audio_duration(theme_file)
                    
                    self.console.print(
                        f"[bold green]✓ SUCCESS[/bold green] [dim]Source:[/dim] {source_name} "
                        f"[dim]|[/dim] {file_size} [dim]|[/dim] {duration}"
                    )
                    
                    # Log success with details
                    self.logger.info(
                        f"Successfully downloaded theme for '{show_name}' from {source_name} "
                        f"(size: {file_size}, duration: {duration})"
                    )
                    
                    self.results["success"] += 1
                    return
                else:
                    self.console.print(f" [red]✗[/red]")
                    self.logger.info(f"Source {source_name} failed for '{show_name}'")
            except OSError as exc:
                # Handle disk space and permission errors during download
                self.console.print(f" [red]✗[/]")
                import errno
                if exc.errno in (errno.ENOSPC, errno.EDQUOT):
                    self.logger.error(
                        f"Disk space error for {show_name}: {str(exc)}",
                        exc_info=True
                    )
                    self.console.print(
                        f"[red]ERROR[/] Disk space error: {str(exc)}"
                    )
                    self.results["failed"] += 1
                    return
                elif exc.errno == errno.EACCES:
                    self.logger.warning(
                        f"Permission denied for {show_name}: {str(exc)}"
                    )
                    self.console.print(
                        f"[yellow]Warning: Permission denied writing to {folder}[/]"
                    )
                    # Continue to next scraper
                else:
                    self.logger.error(
                        f"OS error for {show_name}: {str(exc)}",
                        exc_info=True
                    )
                    if self.verbose:
                        self.console.print(f"    OS Error: {str(exc)}")
            except Exception as exc:
                self.logger.error(
                    f"Unexpected exception in scraper {scraper.get_source_name()} "
                    f"for {show_name}: {str(exc)}",
                    exc_info=True
                )
                self.console.print(f" [red]✗[/]")
                if self.verbose:
                    self.console.print(f"    Error: {str(exc)}")

        self.console.print(f"[bold red]✗ FAILED[/bold red] [dim]No sources found for[/dim] {show_name}")
        self.logger.warning(f"All sources failed for '{show_name}'")
        self.results["failed"] += 1

    def display_summary(self) -> None:
        """Display results summary table."""
        table = Table(title="\n[bold dodger_blue1]Processing Summary[/bold dodger_blue1]", border_style="dodger_blue1")
        table.add_column("Status", style="bold", no_wrap=True)
        table.add_column("Count", justify="right", style="bold")

        table.add_row("✓ Success", f"[bold green]{self.results['success']}[/bold green]")
        table.add_row("⊘ Skipped", f"[yellow]{self.results['skipped']}[/yellow]")
        table.add_row("✗ Failed", f"[bold red]{self.results['failed']}[/bold red]")

        self.console.print(table)

File: src/orchestrator.rs
use crate::config::Config;
use crate::rate_limiter::RateLimiter;
use crate::scrapers::ThemeScraper;
use crate::utils::{get_file_size_formatted, strip_year_from_show_name, validate_show_name};
use anyhow::Result;
use colored::Colorize;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};

/// Configuration for the orchestrator
#[derive(Debug, Clone)]
pub struct OrchestratorConfig {
    /// Overwrite existing theme files
    pub force: bool,
    /// Simulate operations without downloading
    pub dry_run: bool,
    /// Enable verbose/debug logging
    pub verbose: bool,
    /// Network timeout in seconds
    pub timeout: u64,
}

/// Aggregated processing results
#[derive(Debug, Clone, Default)]
pub struct ProcessingResults {
    /// Number of successfully downloaded themes
    pub success: u32,
    /// Number of skipped shows (already have themes)
    pub skipped: u32,
    /// Number of failed shows (all sources failed)
    pub failed: u32,
}

/// Represents a show folder to be processed
#[derive(Debug, Clone)]
pub struct ShowFolder {
    /// Full path to the show folder
    pub path: PathBuf,
    /// Original folder name
    pub name: String,
    /// Search name with year stripped (e.g., "The Simpsons (1989)" -> "The Simpsons")
    pub search_name: String,
}

/// Result of processing a single show
#[derive(Debug, Clone)]
pub enum ShowResult {
    /// Theme was successfully downloaded
    Success {
        source: String,
        file_size: String,
        duration: String,
    },
    /// Show was skipped (already has theme)
    Skipped { reason: String },
    /// All sources failed to find/download theme
    Failed { attempted_sources: Vec<String> },
    /// Dry run mode (no actual download)
    DryRun,
}

/// Main orchestrator that coordinates directory scanning and scraper execution
pub struct Orchestrator {
    /// List of scrapers to try in priority order
    scrapers: Vec<Box<dyn ThemeScraper>>,
    /// Rate limiter to prevent overwhelming sources
    rate_limiter: RateLimiter,
    /// Configuration options
    config: OrchestratorConfig,
    /// Accumulated results
    results: ProcessingResults,
    /// Cancellation flag (set by Ctrl+C handler)
    cancel_flag: Arc<AtomicBool>,
}

impl Orchestrator {
    /// Create a new orchestrator with the given configuration and scrapers
    #[must_use]
    pub fn new(config: OrchestratorConfig, scrapers: Vec<Box<dyn ThemeScraper>>) -> Self {
        Self::with_cancel_flag(config, scrapers, Arc::new(AtomicBool::new(false)))
    }

    /// Create a new orchestrator with a cancellation flag for graceful shutdown
    #[must_use]
    pub fn with_cancel_flag(
        config: OrchestratorConfig,
        scrapers: Vec<Box<dyn ThemeScraper>>,
        cancel_flag: Arc<AtomicBool>,
    ) -> Self {
        let rate_limiter = RateLimiter::new();

        Self {
            scrapers,
            rate_limiter,
            config,
            results: ProcessingResults::default(),
            cancel_flag,
        }
    }

    /// Get the current processing results
    #[must_use]
    pub const fn results(&self) -> &ProcessingResults {
        &self.results
    }

    /// Process a directory containing show folders
    ///
    /// # Errors
    ///
    /// Returns an error if the directory cannot be read or if there are critical I/O failures.
    pub async fn process_directory(&mut self, input_dir: &Path) -> Result<()> {
        // Scan directory for show folders
        let show_folders = Self::scan_directory(input_dir)?;

        if show_folders.is_empty() {
            println!(
                "{}",
                "No series folders found in the input directory.".yellow()
            );
            return Ok(());
        }

        // Display total count
        println!(
            "\n{} {} show folders\n",
            "Found".green().bold(),
            show_folders.len()
        );

        // Process each show folder
        let total = show_folders.len();
        for (index, show_folder) in show_folders.iter().enumerate() {
            // Check for cancellation before each show
            if self.cancel_flag.load(Ordering::SeqCst) {
                tracing::info!("Cancellation requested, stopping after {} shows", index);
                println!(
                    "\n{} Stopped by user after processing {}/{} shows",
                    "⚠".yellow(),
                    index,
                    total
                );
                break;
            }

            let folder_num = index + 1;
            self.process_show(show_folder, folder_num, total).await;
        }

        // Display summary
        self.display_summary();

        Ok(())
    }

    /// Scan directory and return list of show folders
    fn scan_directory(input_dir: &Path) -> Result<Vec<ShowFolder>> {
        tracing::info!("Scanning directory: {}", input_dir.display());

        let mut show_folders = Vec::new();

        let entries = fs::read_dir(input_dir)?;

        for entry in entries {
            let entry = match entry {
                Ok(e) => e,
                Err(err) => {
                    tracing::warn!("Failed to read directory entry: {}", err);
                    eprintln!(
                        "{} Failed to read directory entry: {}",
                        "Warning:".yellow(),
                        err
                    );
                    continue;
                }
            };

            let path = entry.path();

            // Skip if not a directory
            if !path.is_dir() {
                continue;
            }

            // Get folder name
            let name = match path.file_name() {
                Some(n) => n.to_string_lossy().to_string(),
                None => continue,
            };

            // Validate show name
            if !validate_show_name(&name) {
                tracing::warn!("Skipping folder '{}': invalid or too long name", name);
                eprintln!(
                    "{} Skipping folder '{}': invalid or too long name",
                    "Warning:".yellow(),
                    name
                );
                continue;
            }

            // Strip year from name for search
            let search_name = strip_year_from_show_name(&name);

            tracing::debug!("Found show folder: {} (search: {})", name, search_name);

            show_folders.push(ShowFolder {
                path,
                name,
                search_name,
            });
        }

        // Sort by name for reproducible processing order
        show_folders.sort_by(|a, b| a.name.cmp(&b.name));

        tracing::info!("Found {} show folders", show_folders.len());
        Ok(show_folders)
    }

    /// Process a single show folder
    async fn process_show(&mut self, show_folder: &ShowFolder, folder_num: usize, total: usize) {
        tracing::info!(
            "Processing show {}/{}: {} (search name: {})",
            folder_num,
            total,
            show_folder.name,
            show_folder.search_name
        );

        println!(
            "{} {}/{} - {}",
            "Processing".cyan().bold(),
            folder_num,
            total,
            show_folder.name.bold()
        );

        // Check for existing theme
        if let Some(existing_theme) = Self::check_existing_theme(&show_folder.path) {
            if self.config.force {
                // Force mode: delete existing theme
                tracing::info!("Force mode: deleting existing theme: {}", existing_theme);
                println!(
                    "  {} Deleting existing theme: {}",
                    "⚠".yellow(),
                    existing_theme
                );
                if let Err(err) =
                    fs::remove_file(show_folder.path.join("theme-music").join(&existing_theme))
                {
                    tracing::error!("Failed to delete existing theme: {}", err);
                    eprintln!(
                        "  {} Failed to delete existing theme: {}",
                        "Error:".red(),
                        err
                    );
                    self.results.failed += 1;
                    return;
                }
            } else {
                tracing::info!("Skipping - already has theme: {}", existing_theme);
                println!("  {} Already has theme: {}", "⊘".yellow(), existing_theme);
                self.results.skipped += 1;
                return;
            }
        }

        // Try scrapers in order
        for scraper in &self.scrapers {
            let source_name = scraper.source_name();

            tracing::info!("Attempting scraper: {}", source_name);
            println!("  {} Trying {}...", "→".cyan(), source_name);

            // Ensure theme-music directory exists
            let theme_dir = show_folder.path.join("theme-music");
            if !self.config.dry_run && let Err(err) = fs::create_dir_all(&theme_dir) {
                tracing::error!("Failed to create theme-music directory: {}", err);
                eprintln!(
                    "  {} Failed to create theme-music directory: {}",
                    "Error:".red(),
                    err
                );
                self.results.failed += 1;
                return;
            }

            // Apply rate limiting (skip in dry run to speed up)
            if !self.config.dry_run {
                self.rate_limiter.wait_if_needed(source_name).await;
            }

            // Try to download (or simulate in dry run)
            match scraper
                .search_and_download(
                    &show_folder.search_name,
                    &show_folder.path.join("theme-music").join("theme.mp3"),
                    self.config.dry_run,
                )
                .await
            {
                Ok(true) => {
                    // Success (or would succeed in dry run)
                    if self.config.dry_run {
                        tracing::info!("Dry run: Would download from {}", source_name);
                        println!(
                            "  {} Would download from {}",
                            "✓".green().bold(),
                            source_name.green()
                        );
                        self.results.success += 1;
                        return;
                    }

                    // Verify the file was actually written
                    let theme_path = show_folder.path.join("theme-music").join("theme.mp3");
                    if !theme_path.exists() {
                        tracing::warn!(
                            "{} claimed success but theme file not found at: {}",
                            source_name,
                            theme_path.display()
                        );
                        eprintln!(
                            "  {} {} reported success but file not found, trying next source",
                            "⚠".yellow(),
                            source_name
                        );
                        continue;
                    }

                    let file_size = get_file_size_formatted(&theme_path);
                    tracing::info!(
                        "Successfully downloaded from {} ({})",
                        source_name,
                        file_size
                    );
                    println!(
                        "  {} Downloaded from {} ({})",
                        "✓".green().bold(),
                        source_name.green(),
                        file_size
                    );
                    self.results.success += 1;
                    return;
                }
                Ok(false) => {
                    // Not found at this source
                    tracing::info!("Not found at {}", source_name);
                    println!("  {} Not found at {}", "✗".red(), source_name);
                }
                Err(err) => {
                    // Error occurred
                    tracing::error!("Error with {}: {}", source_name, err);
                    eprintln!("  {} Error with {}: {}", "✗".red(), source_name, err);
                }
            }
        }

        // All sources failed
        tracing::warn!("Failed to find theme for: {}", show_folder.name);
        println!("  {} Failed to find theme", "✗".red().bold());
        self.results.failed += 1;
    }

    /// Check if a theme file already exists in the folder
    fn check_existing_theme(folder_path: &Path) -> Option<String> {
        for ext in Config::THEME_EXTENSIONS {
            let theme_file = format!("theme{ext}");
            let theme_path = folder_path.join("theme-music").join(&theme_file);
            tracing::debug!(
                "Checking for existing theme: {} (exists={})",
                theme_path.display(),
                theme_path.exists()
            );
            if theme_path.exists() {
                return Some(theme_file);
            }
        }
        None
    }

    /// Display summary of processing results
    fn display_summary(&self) {
        println!("\n{}", "═".repeat(50));
        println!("{}", "Summary".bold().cyan());
        println!("{}", "═".repeat(50));
        println!(
            "  {} Successful downloads: {}",
            "✓".green(),
            self.results.success
        );
        println!(
            "  {} Skipped (already have theme): {}",
            "⊘".yellow(),
            self.results.skipped
        );
        println!(
            "  {} Failed (not found): {}",
            "✗".red(),
            self.results.failed
        );
        println!("{}", "═".repeat(50));
    }
}
